target/linux/generic/pending-5.15/742-net-dsa-mt7530-use-external-PCS-driver.patch

```diff
@@ -0,0 +1,514 @@
+ }
+ 
+-static int mt7531_sgmii_setup_mode_force(struct mt7530_priv *priv, u32 port,
+-					 phy_interface_t interface)
+-{
+-	u32 val;
+-
+-	if (!mt753x_is_mac_port(port))
+-		return -EINVAL;
+-
+-	mt7530_set(priv, MT7531_QPHY_PWR_STATE_CTRL(port),
+-		   MT7531_SGMII_PHYA_PWD);
+-
+-	val = mt7530_read(priv, MT7531_PHYA_CTRL_SIGNAL3(port));
+-	val &= ~MT7531_RG_TPHY_SPEED_MASK;
+-	/* Setup 2.5 times faster clock for 2.5Gbps data speeds with 10B/8B
+-	 * encoding.
+-	 */
+-	val |= (interface == PHY_INTERFACE_MODE_2500BASEX) ?
+-		MT7531_RG_TPHY_SPEED_3_125G : MT7531_RG_TPHY_SPEED_1_25G;
+-	mt7530_write(priv, MT7531_PHYA_CTRL_SIGNAL3(port), val);
+-
+-	mt7530_clear(priv, MT7531_PCS_CONTROL_1(port), MT7531_SGMII_AN_ENABLE);
+-
+-	/* MT7531 SGMII 1G and 2.5G force mode can only work in full duplex
+-	 * mode, no matter MT7531_SGMII_FORCE_HALF_DUPLEX is set or not.
+-	 */
+-	mt7530_rmw(priv, MT7531_SGMII_MODE(port),
+-		   MT7531_SGMII_IF_MODE_MASK | MT7531_SGMII_REMOTE_FAULT_DIS,
+-		   MT7531_SGMII_FORCE_SPEED_1000);
+-
+-	mt7530_write(priv, MT7531_QPHY_PWR_STATE_CTRL(port), 0);
+-
+-	return 0;
+-}
+-
+-static int mt7531_sgmii_setup_mode_an(struct mt7530_priv *priv, int port,
+-				      phy_interface_t interface)
+-{
+-	if (!mt753x_is_mac_port(port))
+-		return -EINVAL;
+-
+-	mt7530_set(priv, MT7531_QPHY_PWR_STATE_CTRL(port),
+-		   MT7531_SGMII_PHYA_PWD);
+-
+-	mt7530_rmw(priv, MT7531_PHYA_CTRL_SIGNAL3(port),
+-		   MT7531_RG_TPHY_SPEED_MASK, MT7531_RG_TPHY_SPEED_1_25G);
+-
+-	mt7530_set(priv, MT7531_SGMII_MODE(port),
+-		   MT7531_SGMII_REMOTE_FAULT_DIS |
+-		   MT7531_SGMII_SPEED_DUPLEX_AN);
+-
+-	mt7530_rmw(priv, MT7531_PCS_SPEED_ABILITY(port),
+-		   MT7531_SGMII_TX_CONFIG_MASK, 1);
+-
+-	mt7530_set(priv, MT7531_PCS_CONTROL_1(port), MT7531_SGMII_AN_ENABLE);
+-
+-	mt7530_set(priv, MT7531_PCS_CONTROL_1(port), MT7531_SGMII_AN_RESTART);
+-
+-	mt7530_write(priv, MT7531_QPHY_PWR_STATE_CTRL(port), 0);
+-
+-	return 0;
+-}
+-
+-static void mt7531_pcs_an_restart(struct phylink_pcs *pcs)
+-{
+-	struct mt7530_priv *priv = pcs_to_mt753x_pcs(pcs)->priv;
+-	int port = pcs_to_mt753x_pcs(pcs)->port;
+-	u32 val;
+-
+-	/* Only restart AN when AN is enabled */
+-	val = mt7530_read(priv, MT7531_PCS_CONTROL_1(port));
+-	if (val & MT7531_SGMII_AN_ENABLE) {
+-		val |= MT7531_SGMII_AN_RESTART;
+-		mt7530_write(priv, MT7531_PCS_CONTROL_1(port), val);
+-	}
+-}
+-
+ static int
+ mt7531_mac_config(struct dsa_switch *ds, int port, unsigned int mode,
+ 		  phy_interface_t interface)
+@@ -2711,11 +2595,11 @@ mt7531_mac_config(struct dsa_switch *ds,
+ 		phydev = dp->slave->phydev;
+ 		return mt7531_rgmii_setup(priv, port, interface, phydev);
+ 	case PHY_INTERFACE_MODE_SGMII:
+-		return mt7531_sgmii_setup_mode_an(priv, port, interface);
+ 	case PHY_INTERFACE_MODE_NA:
+ 	case PHY_INTERFACE_MODE_1000BASEX:
+ 	case PHY_INTERFACE_MODE_2500BASEX:
+-		return mt7531_sgmii_setup_mode_force(priv, port, interface);
++		/* handled in SGMII PCS driver */
```

Comment:
Sorry but there seems to be a problem here? I can't connect with 1g speed on mt7531 + rtl8221b.

Reply:
@bmork Has been trying this on devices having a 2.5G PHY connected to MT7531, I think it was also RTL8221B, so maybe he can help there. I only got a board with GPY211.

Reply:
I've been reading `drivers/net/phy/realtek.c` and there is nothing there switching the interface mode between SGMII and 2500Base-X. So this phy driver will need more work to inform the system about this, similar to how it is done in `mxl-gpy.c`, there you can see function `gpy_update_interface` taking care of this.

Reply:
Of course I applied my patch first, and `ethtool lan2` looks fine:
![t](https://user-images.githubusercontent.com/42570690/226178960-15b5dda4-c1cd-4aad-8b19-167448586d2c.png)

Reply:
Please dump SerDes related registers of the MT7531:
```
cat /sys/kernel/debug/regmap/mdio-bus:1f-port5/registers
```

Reply:
https://pastebin.com/3naXXqsG

Reply:
Please also check the value of this register of the RealTek PHY.
![serdes-an](https://user-images.githubusercontent.com/9948313/226182209-398f0ba8-2bee-40be-8034-1882a6275789.png)

If SerDes AN is enabled, SerDes link will not come up unless it is enabled also on the other side (ie. MT7531). As SerDes AN (aka. in-band-status) is typically not use for MAC<->PHY connections but mostly for SFP links, you are advised to make sure it is switched off also on the RealTek PHY.

To play with it you can use `mdio-tools`

Reply:
> https://pastebin.com/3naXXqsG
> ```
> 000: 00040140
> ```

This looks like SerDes link is up from perspective of the MT7531 (bit 18 is 1). However, as expected when in-band-status is not in use SGMII autonegotiation is switched off (bit 12 is 0). Hence the same needs to be true on the side of the PHY, ie. MMD device 30, register 0x7582 bit 7 needs to be 0.

Reply:
> ie. MMD device 30, register 0x7582 bit 7 needs to be 0.

Thanks, am I doing this correctly?
Read: `mdio mdio-bus mmd 5:30 raw 0x7582` Output: 0x49f6
Write: `mdio mdio-bus mmd 5:30 raw 0x7582 0x4976`

Reply:
Yes, looks correct.
And indeed bit 7 was originally set (maybe by the bootloader or post-reset default value). Make sure to trigger a re-connect (ie. pull and replug cable, or even plug a 2500Base-T link and then try again with 1000Base-T, so the PHY realizes that it should re-try establishing a link to the MAC.

Reply:
You can try this patch
```diff
diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index 3d99fd6664d7a..3b1f32a471961 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -675,6 +675,17 @@ static int rtl822x_config_aneg(struct phy_device *phydev)
        return __genphy_config_aneg(phydev, ret);
 }
 
+static int rtl8221_config_aneg(struct phy_device *phydev)
+{
+       int ret = 0;
+
+       ret = phy_modify_mmd_changed(phydev, MDIO_MMD_VEND1, 0x7582, BIT(7), 0);
+       if (ret == 1)
+               ret = genphy_soft_reset(phydev);
+
+       return rtl822x_config_aneg(phydev);
+}
+
 static int rtl822x_read_status(struct phy_device *phydev)
 {
        int ret;
@@ -1001,7 +1012,7 @@ static struct phy_driver realtek_drvs[] = {
                PHY_ID_MATCH_EXACT(0x001cc849),
                .name           = "RTL8221B-VB-CG 2.5Gbps PHY",
                .get_features   = rtl822x_get_features,
-               .config_aneg    = rtl822x_config_aneg,
+               .config_aneg    = rtl8221_config_aneg,
                .read_status    = rtl822x_read_status,
                .suspend        = genphy_suspend,
                .resume         = rtlgen_resume,
@@ -1011,7 +1022,7 @@ static struct phy_driver realtek_drvs[] = {
                PHY_ID_MATCH_EXACT(0x001cc84a),
                .name           = "RTL8221B-VM-CG 2.5Gbps PHY",
                .get_features   = rtl822x_get_features,
-               .config_aneg    = rtl822x_config_aneg,
+               .config_aneg    = rtl8221_config_aneg,
                .read_status    = rtl822x_read_status,
                .suspend        = genphy_suspend,
                .resume         = rtlgen_resume,
```

Also let me know which exact PHY you are trying this with.

Reply:
> You can try this patch [PHY_ID(0x001cc849)]

Unfortunately neither using mdio-tools nor this patch can make it work.

Reply:
Please share a complete bootlog and also the exact type of the connected PHY.

Reply:
I assume you are using a RTL8221 variant.
On top of https://github.com/openwrt/openwrt/commit/674ee0bacde4b40ad7bea29698daed3ad242dc57, please additionally apply this patch to make sure the PHY is set to correct 2500Base-X+SGMII mode.

```diff
diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index 3d99fd6664d7a..33cbd0f2fe4c3 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -53,6 +53,26 @@
 						 RTL8201F_ISR_LINK)
 #define RTL8201F_IER				0x13
 
+#define RTL8221_SERDES_OPTION_1			0x697A
+#define RTL8221_SERDES_MODE_MASK		GENMASK(5, 0)
+#define RTL8221_SERDES_MODE_2500BASEX_SGMII	0x0
+#define RTL8221_SERDES_MODE_HISGMII_SGMII	0x1
+#define RTL8221_SERDES_MODE_2500BASEX		0x2
+#define RTL8221_SERDES_MODE_HISGMII		0x3
+#define RTL8221_SERDES_LINK_DOWN_OPTION		BIT(15)
+
+#define RTL8221_SERDES_CONTROL_5		0x7582
+#define RTL8221_FULL_DUPLEX			BIT(2)
+#define RTL8221_RX_PAUSE			BIT(5)
+#define RTL8221_TX_PAUSE			BIT(6)
+#define RTL8221_AN_ENABLE			BIT(7)
+#define RTL8221_PORT_SPEED_MASK			(GENMASK(13, 12) | GENMASK(1, 0))
+#define RTL8221_PORT_SPEED_10M			0x0
+#define RTL8221_PORT_SPEED_100M			0x1
+#define RTL8221_PORT_SPEED_1000M		0x2
+#define RTL8221_PORT_SPEED_2500M		0x1001
+#define RTL8221_PORT_SPEED_2500M_LITE		0x1003
+
 #define RTL8366RB_POWER_SAVE			0x15
 #define RTL8366RB_POWER_SAVE_ON			BIT(12)
 
@@ -675,6 +695,22 @@ static int rtl822x_config_aneg(struct phy_device *phydev)
 	return __genphy_config_aneg(phydev, ret);
 }
 
+static int rtl8221_config_init(struct phy_device *phydev)
+{
+	/* set 2500Base-X+SGMII mode */
+	phy_modify_mmd(phydev, MDIO_MMD_VEND1, RTL8221_SERDES_OPTION_1,
+		       RTL8221_SERDES_MODE_MASK |
+		       RTL8221_SERDES_LINK_DOWN_OPTION,
+		       RTL8221_SERDES_MODE_2500BASEX_SGMII |
+		       RTL8221_SERDES_LINK_DOWN_OPTION);
+
+	/* disable SerDes AN */
+	phy_clear_bits_mmd(phydev, MDIO_MMD_VEND1,
+			   RTL8221_SERDES_CONTROL_5, RTL8221_AN_ENABLE);
+
+	return 0;
+}
+
 static int rtl822x_read_status(struct phy_device *phydev)
 {
 	int ret;
@@ -1002,6 +1038,7 @@ static struct phy_driver realtek_drvs[] = {
 		.name           = "RTL8221B-VB-CG 2.5Gbps PHY",
 		.get_features   = rtl822x_get_features,
 		.config_aneg    = rtl822x_config_aneg,
+		.config_init    = rtl8221_config_init,
 		.read_status    = rtl822x_read_status,
 		.suspend        = genphy_suspend,
 		.resume         = rtlgen_resume,
@@ -1012,6 +1049,7 @@ static struct phy_driver realtek_drvs[] = {
 		.name           = "RTL8221B-VM-CG 2.5Gbps PHY",
 		.get_features   = rtl822x_get_features,
 		.config_aneg    = rtl822x_config_aneg,
+		.config_init    = rtl8221_config_init,
 		.read_status    = rtl822x_read_status,
 		.suspend        = genphy_suspend,
 		.resume         = rtlgen_resume,
```

I cannot test because I don't have any device using this PHY at hand, but you get the idea. Also I don't know if we really need to set `RTL8221_SERDES_LINK_DOWN_OPTION`, the datasheet is a bit ambiguous when it comes to that bit, so you can try with setting it to 1 and to 0 and see if that helps.

Generally we are interested in registers
 - 30.0x697A SERDES OPTION
 - 30.0x6A04 SERDES Control 1 Register
 - 30.0x7580 SERDES Control 3 Register
 - 30.0x7581 SERDES Control 4 Register
 - 30.0x7582 SERDES Control 5 Register
 - 30.0x7587 SERDES Control 6 Register
 - 30.0x758B SERDES Control 7 Register
 - 30.0x758C SERDES Control 8 Register
 - 30.0x758D SERDES Control 9 Register

So in case the patch above doesn't help to bring the link up, please dump those registers using `mdio` tool.

Reply:
[RTL8221B-VB-CG (0x001cc849)]
https://pastebin.com/LhUtr5Qb

Reply:
On the current source of openwrt master, the rtl8221b phy connected to mt7531 can work normally (with my patch). It was broken after applying the changes in this pull request, and the patch above doesn't fix it. Sorry I don't have more time to test it.

Reply:
@dangowrt  
I have a board consisting of 7981, two MT7531 chips, and two RTL8221B_PHY chips. I believe I am capable of conducting the test, but unfortunately, the RTL8221B_PHY chips are not currently functioning. Do you have any suggestions for troubleshooting the issue?

Reply:
@ptpt52 I believe the problem is related to the RTL8221B PHYs not being initialized correctly. Apart from the patch by @aiamadeus we will need to introduce a `.config_init` function which sets the PHY to operate in SGMII mode without in-band-status for 10M/100M/1000M and switch to 2500Base-X for 2500M. I've (blindly because I don't have the hardware yet) created such a `.config_init` function, but maybe other things are still wrong there. However, it should provide a good starting point:
https://github.com/openwrt/openwrt/pull/12202#discussion_r1141632220

As @aiamadeus has already reported that the patch doesn't work as-is, my next try would be to remove `RTL8221_SERDES_LINK_DOWN_OPTION` and re-try without that option set. The datasheets you can find also mention that there is a "SerDes application notes" document which I don't have, and probably that document mentions the correct sequence how to initialize the PHY for any possible modes....